Fixed composedClassLists breaking order

otaviomad opened the pull request:
Closes #1430 

This PR achieves this by doing the following:

- Changing scanModule to be a recursive function that always keeps the dependencies last in the array order
- Building up the provided classlist based on the module id, so that it's always in order

The way that I did these things might not be to the Vanilla Extract's maintainers standard, so consider this a draft or a suggestion. I have no clue if this breaks something else as well, this is just me figuring out how this issue could be fixed.

otaviomad commented:
I see that I broke some tests by literally changing the orders. If they are not issues, I can update them to match the new order.

askoufis commented:
@otaviomad Sorry for taking so long to get to this PR, I kept getting stuck on reproducing the issue in a test. I've pushed the test to a branch - you can see the incorrect class name [here](https://github.com/vanilla-extract-css/vanilla-extract/compare/class-composition-bug-test?expand=1#diff-bfde3d87ff463cc28a87bb59fb75bb0258037b96a7264034d5ff8724c8f4585cR343). However, when I apply your fix to the compiler, it doesn't remove the unwanted `.base` class, but rather just moves it.

Hopefully the unit test can help you debug the issue. Thanks for looking into this, it's very much appreciated.

otaviomad commented:
@askoufis Impressive work!

I am on vacation at the moment and will only be able to look at it in two weeks. I'll pick it up as soon as I'm back :)

askoufis commented:
> @askoufis Impressive work!
> 
> I am on vacation at the moment and will only be able to look at it in two weeks. I'll pick it up as soon as I'm back :)

No worries, don't want to rush you. Enjoy your vacation!

otaviomad commented:
I believe I have fixed it. The problem arose from having some confusion on my `scanModule` re-implementation. 

The way that it works is that it _should_ return cssDeps as an array with the order being: 
nested modules -> shallow dep modules -> the actual scanned module

What was happening is that, to help push the nested modules to the beginning of the array, I would use a Set and for every occurrence of a module, I would delete it and add it again to the Set. However, this would rely on the nested modules actually being _ordered to the end_ of the array, _not the beginning_, which is what scanModule actually returned (which I did to keep the API similar to the previous version without having to reverse it in place on the compiler). So the fix was just to reverse the nested scanModule result to get the proper order that we wanted :)

otaviomad commented:
It seems playwright is failing, probably due to ordering stuffs. I dunno how I'd fix this one though. Any help?

renrizzolo commented:
> It seems playwright is failing, probably due to ordering stuffs. I dunno how I'd fix this one though. Any help?

assuming the new output is what you expect, you can locally change `updateSnapshots` in `playwright.config.ts` to `"all"`, then run `test:playwright`. 
Then verify/commit the changed snapshots and revert the playwright config change.

otaviomad commented:
It's done :) 
@askoufis may we get a formal review?

smitev commented:
Any progress on this one?
Seems like this will fix one of the issues we have, too.
I hope this will be merged in soon.

askoufis commented on tests/e2e/thirdparty.playwright.ts-snapshots/thirdparty-esbuild-next--development.css:
```diff
@@ -1,11 +1,11 @@
-  --depColor__vg1esd0: green;
-  background-color: var(--depColor__vg1esd0);
-}
 .styles_depdepBlock__7q0fkw1 {
   --depdepColor__7q0fkw0: blue;
   background-color: var(--depdepColor__7q0fkw0);
 }
+.styles_depBlock__vg1esd1 {
+  --depColor__vg1esd0: green;
+  background-color: var(--depColor__vg1esd0);
+}
```
I don't think this ordering is correct. Looking at the import order of styles in the `thirdparty` fixture (see [here](https://github.com/vanilla-extract-css/vanilla-extract/blob/9a0c6898e02ffe54c8455110ac0a7adecb334acc/fixtures/thirdparty/thirdparty-dep/index.mjs)), I would expect the `depBlock` style to be declared before the `depdepBlock` style.

otaviomad replied:
I think I agree. I'll have a look at it to figure out why the order changed here.

otaviomad commented:
I _believe_ I fixed everything now. I changed the way it works so that there isn't a weird cache argument being mutated and no longer reversing it twice (which was the culprit, in the end). After hitting my head in the wall for a bit, I realized that "local" dependencies had to be kept in order as well as having the nested deps further up in the dep array. I changed the implementation to use an array, which will keep copies of the nested dependency ids unshifted to the beginning, only to be filtered out by creating a set later, which will make the array ids unique. 

Hopefully this is good enough now :)

otaviomad commented:
Turns out I was thinking about this wrong. It was never about unshifting or changing orders - the problem lied in the fact that scanModule ran sequentially - thus breaking the self-insertion of css modules as dependencies. Changing it to a recursive function was the fix from the get go, no reversing or breaking stuff needed to begin with. This allowed me to simplify some things - I think it's pretty good now and shouldn't break behaviour anywhere :)

askoufis commented:
Thanks for persevering on this @otaviomad. The implementation looks good to me. However, I wonder if the effect on CSS output is significant enough to warrant a major version for any consuming packages of the compiler (vite and esbuild-next). Thoughts?

otaviomad commented:
@askoufis I believe these are minor changes. No majors required.

askoufis commented:
> @askoufis I believe these are minor changes. No majors required.

I guess my concern was that someone could have written some extra styles to combat incorrect styles generated by this bug (which they don't necessarily know about). By fixing the CSS output, it _could_ result in those extra styles causing unwanted styling. I think it's likely a very rare edge case though, so I'm happy to merge this as a patch.

askoufis commented on .changeset/late-icons-protect.md:
```diff
@@ -0,0 +1,5 @@
+"@vanilla-extract/integration": patch
+---
+
+Fixed composedClassLists breaking order
```
```suggestion
Fixes a bug in the `vite-node` compiler resulting in incorrectly ordered class compositions in certain cases
```
